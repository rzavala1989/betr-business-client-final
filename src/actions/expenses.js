// import uuid from 'uuid';  /* <---NOT NEEDED */

import {API_BASE_URL} from "../config";
import {normalizeResponseErrors} from './utils';

export const FETCH_EXPENSE_DATA_SUCCESS = "FETCH_EXPENSE_DATA_SUCCESS";
export const fetchExpenseDataSuccess = data => ({
  type: FETCH_EXPENSE_DATA_SUCCESS,
  data
});

export const FETCH_EXPENSE_DATA_ERROR = "FETCH_EXPENSE_DATA_ERROR";
export const fetchExpenseDataError = error => ({
  type: FETCH_EXPENSE_DATA_ERROR,
  error
});

export const fetchExpenseData = () => (dispatch, getState) => {
  const authToken = getState().auth.authToken;
  return fetch(`${API_BASE_URL}/expenses`, {
    method: "GET",
    headers: {
      // Provide our auth token as credentials
      Authorization: `Bearer ${authToken}`
    }
  })
    .then(res => normalizeResponseErrors(res))
    .then(res => res.json())
    .then(data => dispatch(fetchExpenseDataSuccess(data)))
    .catch(err => {
      dispatch(fetchExpenseDataError(err));
    });
};

export const POST_EXPENSE_DATA_SUCCESS = "POST_EXPENSE_DATA_SUCCESS";
export const postExpenseDataSuccess = data => ({
  type: POST_EXPENSE_DATA_SUCCESS,
  data
});

export const POST_EXPENSE_DATA_ERROR = "POST_EXPENSE_DATA_ERROR";
export const postExpenseDataError = error => ({
  type: POST_EXPENSE_DATA_ERROR,
  error
});

export const postExpense = expense => (dispatch, getState) => {
  const authToken = getState().auth.authToken;
  return fetch(`${API_BASE_URL}/expenses`, {
    method: "POST",
    headers: {
      // Provide our auth token as credentials
      Authorization: `Bearer ${authToken}`,
      "Content-type": "application/json"
    },
    body: JSON.stringify(expense)
  })
    .then(res => normalizeResponseErrors(res))
    .then(res => res.json())
    .then(data => dispatch(postExpenseDataSuccess(data)))
    .catch(err => {
      dispatch(postExpenseDataError(err));
    });
};

export const EDIT_EXPENSE_SUCCESS = "EDIT_EXPENSE_SUCCESS";
export const editExpenseSuccess = (id, updates) => ({
  type: EDIT_EXPENSE_SUCCESS,
  id,
  updates
});

export const EDIT_EXPENSE_ERROR = "EDIT_EXPENSE_ERROR";
export const editExpenseError = error => ({
  type: EDIT_EXPENSE_ERROR,
  error
});

export const editExpense = (id, updates) => (dispatch, getState) => {
  updates.id = id;
  const authToken = getState().auth.authToken;
  fetch(`${API_BASE_URL}/expenses/${id}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${authToken}`
    },
    body: JSON.stringify(updates)
  })
    .then(res => dispatch(editExpenseSuccess(updates)))
    .catch(err => {
      dispatch(editExpenseError(err));
    });
};

export const DELETE_EXPENSE_SUCCESS =
  "DELETE_EXPENSE_SUCCESS";
export const deleteExpenseSuccess = ({id}) => ({
  type: DELETE_EXPENSE_SUCCESS,
  id
});

export const DELETE_EXPENSE_ERROR = "DELETE_EXPENSE_ERROR";
export const deleteExpenseError = error => ({
  type: DELETE_EXPENSE_ERROR,
  error
});

export const deleteExpense = ({id}) => (dispatch, getState) => {
  const authToken = getState().auth.authToken;
  return fetch(`${API_BASE_URL}/expenses/${id}`, {
    method: "DELETE",
    headers: {
      // Provide our auth token as credentials
      Authorization: `Bearer ${authToken}`
    }
  }).then(res => dispatch(deleteExpenseSuccess({id})));
};

// export const SET_SELECTED_EXPENSE = "SET_SELECTED_EXPENSE";
// export const setSelectedExpense = (expense, username) => ({
//   type: SET_SELECTED_EXPENSE,
//   expense,
//   username
// });

// ________________________________

// ADD_EXPENSE
// export const addExpense = (
//   {
//     description = '',
//     note = '',
//     amount = 0,
//     createdAt = 0
//   } = {}
// ) => ({
//   type: 'ADD_EXPENSE',
//   expense: {
//     description,
//     note,
//     amount,
//     createdAt
//   }
// });

// REMOVE_EXPENSE
// export const removeExpense = ({ id } = {}) => ({
//   type: 'REMOVE_EXPENSE',
//   id
// });

// EDIT_EXPENSE
// export const editExpense = (id, updates) => ({
//   type: 'EDIT_EXPENSE',
//   id,
//   updates
// });
