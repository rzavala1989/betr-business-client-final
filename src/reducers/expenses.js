// Expenses Reducer
import {
  FETCH_EXPENSE_DATA_SUCCESS,
  FETCH_EXPENSE_DATA_ERROR,
  POST_EXPENSE_DATA_SUCCESS,
  POST_EXPENSE_DATA_ERROR,
  EDIT_EXPENSE_SUCCESS,
  EDIT_EXPENSE_ERROR,
  DELETE_EXPENSE_SUCCESS,
  DELETE_EXPENSE_ERROR
} from "../actions/expenses";

const initialState = {
  expenses: []
};

export default function reducer(state = initialState, action) {
  if (action.type === FETCH_EXPENSE_DATA_SUCCESS) {
    return Object.assign({}, state, { expenses: action.data, error: null });
  }
  if (action.type === FETCH_EXPENSE_DATA_ERROR) {
    return Object.assign({}, state, { error: action.error });
  }
  if (action.type === POST_EXPENSE_DATA_SUCCESS) {
    return Object.assign({}, state, {
      expenses: [...state.expenses, action.data],
      error: null
    });
  }
  if (action.type === POST_EXPENSE_DATA_ERROR) {
    return Object.assign({}, state, { error: action.error });
  }
  if (action.type === EDIT_EXPENSE_SUCCESS) {
    return Object.assign({}, state, {
      expenses: state.expenses.map(
        expense =>
          expense.id === action.id ? action.updates: expense
      )
    });
  }
  if (action.type === EDIT_EXPENSE_ERROR) {
    return Object.assign({}, state, { expenses: action.error });
  }
  if (action.type === DELETE_EXPENSE_SUCCESS) {
    return Object.assign({}, state, {
      ...state.expenses.slice(0, action.index),
      ...state.expenses.slice(action.index + 1)
    });
  }
  if (action.type === DELETE_EXPENSE_ERROR) {
    return Object.assign({}, state, { error: action.error });
  }
  return state;
}

