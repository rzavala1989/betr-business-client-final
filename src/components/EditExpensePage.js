import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, deleteExpense } from '../actions/expenses';


export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    console.log(expense)
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onRemove = () => {
    this.props.deleteExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };
  render() {
    return (
      <div className="edit-expense-page">
      <div className="bg-edit-expense-page"></div>
        <div className="page-header">
          <div className="container">
            <h1 className="page-header__title">Edit Your Expense</h1>
          </div>
        </div>
        <div className="container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button className="button--remove" onClick={this.onRemove}>Remove Expense Forever</button>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  deleteExpense: (data) => dispatch(deleteExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);







// const EditExpensePage = (props) => {
//   return (
//     <div className="edit-expense-page">
//       <div className="bg-edit-expense-page"></div>
//         <div className="page-header">
//           <div className="container">
//             <h1 className="page-header__title">Edit Your Expense</h1>
//           </div>
//         </div>
//         <div className="container">
//           <ExpenseForm
//             expense={props.expense}
//             onSubmit={(expense) => {
//               props.dispatch(editExpense(props.expense.id, expense));
//               props.history.push('/');
//             }}
//           />
//           <button 
//             className="button--remove" 
//             onClick={() => {
//               props.dispatch(deleteExpense({ id: props.expense.id }));
//               props.history.push('/');
//             }}>Remove Expense Forever</button>
//         </div>
//       </div>
//   )
// };

// const mapStateToProps = (state, props) => ({
//   expense: state.expenses.expenses.find((expense) => expense.id === props.match.params.id)
// });

// export default connect(mapStateToProps)(EditExpensePage);


