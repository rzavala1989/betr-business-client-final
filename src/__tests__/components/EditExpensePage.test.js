import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../../test-data/expenses';
import EditExpensePage from '../../components/EditExpensePage';
import {Provider} from 'react-redux';

let editExpense, removeExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <Provider><EditExpensePage
      editExpense={editExpense}
      removeExpense={removeExpense}
      history={history}
      expense={expenses[2]}
    /></Provider>
  );
});


test('should render EditExpensePage', () => {
  expect(wrapper).toMatchSnapshot();
});
