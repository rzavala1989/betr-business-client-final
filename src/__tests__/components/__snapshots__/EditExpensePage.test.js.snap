// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider>
    <Connect(EditExpensePage)
        editExpense={[Function]}
        expense={
            Object {
                "amount": 4500,
                "createdAt": 345600000,
                "description": "Credit Card",
                "id": "3",
                "note": "",
              }
        }
        history={
            Object {
                "push": [Function],
              }
        }
        removeExpense={[Function]}
    />
</Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "editExpense": [Function],
      "expense": Object {
        "amount": 4500,
        "createdAt": 345600000,
        "description": "Credit Card",
        "id": "3",
        "note": "",
      },
      "history": Object {
        "push": [Function],
      },
      "removeExpense": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "editExpense": [Function],
        "expense": Object {
          "amount": 4500,
          "createdAt": 345600000,
          "description": "Credit Card",
          "id": "3",
          "note": "",
        },
        "history": Object {
          "push": [Function],
        },
        "removeExpense": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": true,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): Object {
    "store": undefined,
    "storeSubscription": null,
  },
}
`;
